repository/src/main/java/com/epam/esm/repository/impl/CertificateRepositoryImpl.java

```diff
@@ -0,0 +1,211 @@
+            %s;
+            """;
+
+    private static final String SELECT_CERTIFICATE_BY_ID = """
+            SELECT ID, NAME, DESCRIPTION, PRICE, DURATION, CREATE_DATE, LAST_UPDATE_DATE FROM gift_certificate
+            WHERE id = :id;
+            """;
+
+    private static final String INSERT_TAG_TO_CERTIFICATE = """
+            INSERT INTO gift_certificate_has_tag (gift_certificate_id, tag_id)
+            VALUES (:id, :tag_id);
+            """;
+
+    private static final String DELETE_TAG_FROM_CERTIFICATE = """
+            DELETE FROM gift_certificate_has_tag
+            WHERE gift_certificate_id = :gift_certificate_id AND tag_id = :tag_id;
+            """;
+
+    private static final String INSERT_CERTIFICATE = """
+            INSERT INTO gift_certificate (name, description, price, duration, create_date, last_update_date)
+            VALUES (:name, :description, :price, :duration, :create_date, :last_update_date)
+            """;
+
+    private static final String UPDATE_CERTIFICATE = """
+            UPDATE gift_certificate
+            SET name = :name, description = :description, price = :price, duration = :duration, last_update_date = :last_update_date
+            WHERE id = :id;
+            """;
+
+    private static final String DELETE_CERTIFICATE = """
+            DELETE FROM gift_certificate
+            WHERE id = :id;
+            """;
+
+    private final RowMapper<Certificate> rowMapper;
+    private final NamedParameterJdbcTemplate namedParameterJdbcTemplate;
+
+    public CertificateRepositoryImpl(DataSource dataSource, RowMapper<Certificate> rowMapper) {
+        this.namedParameterJdbcTemplate = new NamedParameterJdbcTemplate(dataSource);
+        this.rowMapper = rowMapper;
+    }
+
+    @Override
+    public List<Certificate> find(String tagName, String certificateName, String certificateDescription
+            , SortType sortByName, SortType sortByCreateDate) {
+
+        MapSqlParameterSource parameters = new MapSqlParameterSource();
+        List<String> findBy = new ArrayList<>();
+        List<String> orderBy = new ArrayList<>();
+
+        if (!tagName.isEmpty()) {
+            findBy.add("tag.name = :" + TAG_NAME_PARAMETER);
+            parameters.addValue(TAG_NAME_PARAMETER, tagName);
+        }
+
+        if (!certificateName.isEmpty()) {
+            findBy.add("gift_certificate.name = :" + NAME_PARAMETER);
+            parameters.addValue(NAME_PARAMETER, certificateName);
+        }
+
+        if (!certificateDescription.isEmpty()) {
+            findBy.add("gift_certificate.description = :" + DESCRIPTION_PARAMETER);
+            parameters.addValue(DESCRIPTION_PARAMETER, certificateDescription);
+        }
+
+        if (sortByName != null) {
+            orderBy.add("gift_certificate.name " + sortByName.name());
+        }
+
+        if (sortByCreateDate != null) {
+            orderBy.add("gift_certificate.create_date " + sortByCreateDate.name());
+        }

```

Comment:
In my opinion, the logic for detecting parameters for building a query should not be here.